ecs-agent/tmds/server.go

```diff
@@ -0,0 +1,155 @@
+// copyright amazon.com inc. or its affiliates. all rights reserved.
+//
+// licensed under the apache license, version 2.0 (the "license"). you may
+// not use this file except in compliance with the license. a copy of the
+// license is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. this file is distributed
+// on an "as is" basis, without warranties or conditions of any kind, either
+// express or implied. see the license for the specific language governing
+// permissions and limitations under the license.
+package tmds
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/didip/tollbooth"
+	"github.com/gorilla/mux"
+
+	"github.com/aws/amazon-ecs-agent/ecs-agent/logger/audit"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/logger/audit/request"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/tmds/logging"
+	muxutils "github.com/aws/amazon-ecs-agent/ecs-agent/tmds/utils/mux"
+)
+
+const (
+	// TMDS IP and port
+	IPv4 = "127.0.0.1"
+	Port = 51679
+)
+
+// IPv4 address for TMDS
+func AddressIPv4() string {
+	return fmt.Sprintf("%s:%d", IPv4, Port)
+}
+
+// Configuration for TMDS
+type Config struct {
+	listenAddress   string        // http server listen address
+	readTimeout     time.Duration // http server read timeout
+	writeTimeout    time.Duration // http server write timeout
+	steadyStateRate float64       // steady request rate limit
+	burstRate       int           // burst request rate limit
+	router          *mux.Router   // router with routes configured
+}
+
+// Function type for updating TMDS config
+type ConfigOpt func(*Config) error
+
+// Set TMDS listen address
+func WithListenAddress(listenAddr string) ConfigOpt {
+	return func(c *Config) error {
```

Comment:
do these setters need a return type if they always return nil?

Reply:
Good point. I agree the error in the return type is redundant. I will change this.

Reply:
Updated